Improve Entry placeholder resolution

c4rlo opened the pull request:
After resolving placeholders in entry attributes (e.g. `{URL}`), previously the code would do it all over again if anything had changed, multiple times up to the recursion limit. This would have the effect of applying a much greater recursion limit, which is confusing and unnecessary, and probably undesired.

Also, optimise and improve resolution of multiple placeholders by (a) being consistent about which placeholder substring to replace (i.e. the one found by the regex, as opposed to possibly one just introduced through a previous substitution), and (b) building the replacement string from scratch rather than doing successive string replacements.

Together, the above two changes are enough to fix #7158 in my testing.

While I was at it, I made some further improvements to the entry-related code:

- Placeholder resolution: remove unnecessary special casing for self-referential placeholders (these are taken care of by existing recursion depth limit).
- `Entry::size()`: when computing tag size, use the same delimiter set as in other places in the code.
- Factor tag delimiter set regex out into global constant.
- Move some constants from being public static data members of `class Entry` to being local to `Entry.cpp` (in an anonymous namespace).
- Migrate some `QRegEx` instances to `QRegularExpression`, the modern alternative.
- Miscellaneous minor code cleanups.

Also fixes #1741 

## Testing strategy

No new functionality. Existing tests look sufficiently comprehensive, and they continue to pass.

## Type of change
- ✅ Bug fix (non-breaking change that fixes an issue)
- ✅ Refactor (significant modification to existing code)

c4rlo committed:
Entry placeholder resolution: don't overdo it

After resolving placeholders, previously the code would do it all over
again if anything had changed, multiple times up to the recursion limit.
This would have the effect of applying a much greater recursion limit,
which is confusing and unnecessary, and probably undesired.

c4rlo committed:
Entry tweaks and minor refactoring

- Entry::size(): when computing tag size, use same delimiter set as in
  other places in the code
- Factor tag delimiter set regex out into global constant
- Placeholder resolution: remove unnecessary special casing for
  self-referential placeholders (these are taken care of by existing
  recursion depth limit)
- Placeholder resolution: less wasteful string building loop
- Move some constants from being public static data members of Entry to
  being local to Entry.cpp (in anonymous namespace)
- Migrate some QRegEx instances to QRegularExpression, the modern
  alternative
- Miscellanous minor code cleanups

droidmonkey commented:
Oh awesome, I have been meaning to clean this code for years.... its a mess. 

Can you also look into the "you've reached maximum depth of replacement" that gets thrown for some special sequences.
https://github.com/keepassxreboot/keepassxc/issues/1741

There are some other long standing issues: https://github.com/keepassxreboot/keepassxc/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22feature%3A+Reference%2FPlaceholders%22

One thing I really want to move towards, this is another PR, is to swap the functions of Entry. Basically if you call `Entry::Password` for example, it will return the fully baked result with references resolved. Calling `Entry::rawPassword` or forcing to `Entry::getAttribute(Entry::PASSWORD)` would return the raw value. You also remove all the setter functions and force use of `Entry::setAttribute(...)`. This wouldn't affect writing to databases since this is handled at the attribute level already. This moves us closer to a proper data model.

droidmonkey commented:
Just checked and the fixes here do not fix the findings in #1741

Setting an entry with a password of `sdss{fdfdf}sdd` is enough to trigger the issue. Just show the password in the preview panel.

c4rlo committed:
Entry: fix hitting recursion limit with {braces}

When encountering a {brace-enclosed} substring, the placeholder
resolution logic would previously keep recursing until it hit the
recursion depth limit (currently 10). This would lead to "Maximum depth
of replacement has been reached" messages, and was also wasting CPU
cycles.

Fixes #1741.

c4rlo commented:
My latest commit here should fix #1741.

droidmonkey commented:
Hmmmm, this isn't a desirable result.

![image](https://github.com/keepassxreboot/keepassxc/assets/2809491/5bfaecfc-2ac5-452e-b7d5-e7fa78fc7d57)

![image](https://github.com/keepassxreboot/keepassxc/assets/2809491/e1684114-d833-4c7e-8e0e-5b811d41784b)

However this is not a introduced behavior by this PR, it happens in 2.7.8 as well.

droidmonkey committed:
Remove this-> pointers
